src/manager/manager.c

```diff
@@ -252,9 +254,191 @@ static int manager_method_ping(sd_bus_message *m, UNUSED void *userdata, UNUSED
+                        }
+
+                        r = sd_bus_message_close_container(reply);
+                        if (r < 0) {
+                                return r;
+                        }
+                        r = sd_bus_message_exit_container(m);
+                        if (r < 0) {
+                                return r;
+                        }
+                }
+
+                r = sd_bus_message_exit_container(m);
+                if (r < 0) {
+                        return r;
+                }
+        }
+
+        r = sd_bus_message_close_container(reply);
+        if (r < 0) {
+                return r;
+        }
+
+        return 0;
+}
+
+
+static void manager_method_list_units_done(ListUnitRequest *req) {
+        /* All sub-requests are done, collect results and freee when done */
+        _cleanup_list_unit_request_ ListUnitRequest *free_me = req;
+
+        _cleanup_sd_bus_message_ sd_bus_message *reply = NULL;
+        int r = sd_bus_message_new_method_return(req->request_message, &reply);
+        if (r >= 0) {
+                r = manager_method_list_units_encode_reply(req, reply);
+        }
+        if (r < 0) {
+                sd_bus_reply_method_errorf(req->request_message, SD_BUS_ERROR_FAILED, "Internal error");
+                return;
+        }
+
+        r = sd_bus_message_send(reply);
+        if (r < 0) {
+                fprintf(stderr, "Failed to send reply: %s\n", strerror(-r));
+                return;
+        }
+}
+
+static void manager_method_list_units_maybe_done(ListUnitRequest *req) {
+        if (req->n_done == req->n_sub) {
+                manager_method_list_units_done(req);
+        }
+}
+
+
+static int manager_list_units_callback(
+                AgentRequest *agent_req, UNUSED sd_bus_message *m, UNUSED sd_bus_error *ret_error) {
+        ListUnitRequest *req = agent_req->userdata;
+        int i = 0;
+
+        for (int i = 0; i < req->n_sub; i++) {
+                if (req->sub[i].agent_req == agent_req) {
+                        break;
+                }
+        }
+
+        assert(i != req->n_sub); /* we should have found the sub request */
+
+        req->sub[i].m = sd_bus_message_ref(m);
+        req->n_done++;
+
+        manager_method_list_units_maybe_done(req);
+
+        return 0;
+}
+
+static int manager_method_list_units(sd_bus_message *m, void *userdata, UNUSED sd_bus_error *ret_error) {
+        Manager *manager = userdata;
+        ListUnitRequest *req = NULL;
+        Node *node = NULL;
+
+        req = malloc0(sizeof(*req) + sizeof(req->sub[0]) * manager->n_nodes);
+        if (req == NULL) {
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Internal error");
+        }
+        req->request_message = sd_bus_message_ref(m);
+
+        int i = 0;
+        LIST_FOREACH(nodes, node, manager->nodes) {
+                _cleanup_agent_request_ AgentRequest *agent_req = node_request_list_units(
```

Comment:
I believe this ```_cleanup_agent_request_``` can be removed because the pointer to AgentRequest was already stolen inside ```node_request_list_units()``` and the memory is freed later in ```list_unit_request_free()```.

Reply:
It all depends on your point of view. Neither of the two is wrong. The ownership of a ref is transferred from the return of node_request_list_units() into the local variable agent_req, and then transferred ot the sub[].agent_req location (via pointer_steal). 

Since this *always* happens we could leave out the cleanup of the agent_req to make the code simpler. On the other hand, it would be less explicit about how we track ownership, and if any code changes later that makes the assignment of agent_req not always happen, then we would need to add this back. For example, if we introduce another call site between the two that could fail with an error.

Reply:
Cool, makes sense!